### ISSUE TITLE
Databinding discrepancy between post/get on complex type

### ISSUE THREAD

**Issue by shoe788** [aspnet/Mvc#6361]
I'm trying to databind some values with the format

```
columns[0][data]=id
columns[0][name]=id
columns[1][data]=name
columns[1][name]=name
...
```

When I post with this format to an action with the signature

`public async Task<IActionResult> Test(TestBind model)`

With my model as

```
public class TestBind
{
    public IEnumerable<TestBindInner> Columns { get; set; }
}

public class TestBindInner
{
    public string Data { get; set; }

    public string Name { get; set; }
}
```

The data is being bound correctly on a post but not a get. Results pic below.

![capture](https://cloud.githubusercontent.com/assets/2416676/26793965/86e88b7e-49e5-11e7-92ef-b2641d686aea.PNG)
![capture1](https://cloud.githubusercontent.com/assets/2416676/26793966/86eb2bc2-49e5-11e7-8976-a24159cd18cb.PNG)

Why is there a problem with this format for a get but not a post?

**Comment by rynowak** [comment]
How are you sending this data with a `GET`? That looks like form data and `GET` doesn't have a body.

**Comment by shoe788** [comment]
In the headers, for example

`GET /api/Test?columns%5B0%5D%5Bdata%5D=id&columns%5B0%5D%5Bname%5D=id`

**Comment by shoe788** [comment]
If I create a custom model binder I can see that the ValueProvider on the ModelBindingContext is giving me data

![capture3](https://cloud.githubusercontent.com/assets/2416676/26794422/39647438-49e7-11e7-9bef-63e25943095c.PNG)

**Comment by rynowak** [comment]
So what's happening here is that we have a value provider for form data that uses the old jQuery syntax:
https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/JQueryFormValueProvider.cs

This is just enabled for form data.

So you have two options:
1. Add your own value provider factory for jquery syntax + query string
2. Use a key format like `columns[0].name` which will work everywhere

**Comment by rynowak** [comment]
Sorry did not mean to close

**Comment by shoe788** [comment]
I do see that for a `GET` the JQuery value provider isn't in the ModelBindingContext. However, I can still use `GetValue` on the QueryStringValueProvider and get the data, as you can see from my pics above.

Is the default model binder skipping those values?

**Comment by rynowak** [comment]
hmm I'll take a second look and see if I can reproduce that part.

**Comment by rynowak** [comment]
> Is the default model binder skipping those values?

Yes. What happens is that model binding is looking for the key `columns[0].name`. The jQuery form value provider will *translate* all of the incoming data from `[..]` as a property delimiter to `.`.

So in the case of the form post you'd see both `columns[0].name` **and** `columns[0][name]` in the value provider. MVC only looks for the former.

**Comment by shoe788** [comment]
Interesting, thank you. Is there a way to inject the JQuery value provider into the ModelBindingContext for this case? In `MvcOptions` I only see ways to add ValueProviders and ModelBinders

**Comment by rynowak** [comment]
> Interesting, thank you. Is there a way to inject the JQuery value provider into the ModelBindingContext for this case?

Sorry I don't really understand what you're asking. Can you provide an example or be more specific?

>  In MvcOptions I only see ways to add ValueProviders and ModelBinders

`MvcOptions` lets you add instances of `IValueProviderFactory` and `IModelBinderProvider`. These are ultimately ways of customizing the set of value providers and model binders.

Resource filters can also modify the set of `IValueProviderFactory` instances that will be used.

**Comment by shoe788** [comment]
> Sorry I don't really understand what you're asking. Can you provide an example or be more specific?

The implementation signature of `IModelBinder` looks like this

`Task BindModelAsync(ModelBindingContext bindingContext)`

If I want to reuse the JQuery value provider and the Default model binders then I need a way to inject the JQuery value provider into the bindingContext so that the model binding can query and find the right values. Are resource filters the way to accomplish that or am I stuck making my own model binder?

**Comment by rynowak** [comment]
You should make a combination of these two classes:
https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/JQueryFormValueProviderFactory.cs
https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/QueryStringValueProviderFactory.cs

And register that in `MvcOptions.ValueProviderFactories` as the last entry. This means that you'll accept the jQuery syntax in the query string, anywhere in your app.

-----

Basically what you need to solve that [this](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/JQueryFormValueProviderFactory.cs#L58) line of code is going to operate on the form data, you want to make it operate on the query string data.

---

We've been discussing this internally and we're likely going to just enable this by default in the future, but this should get you going.

**Comment by shoe788** [comment]
>And register that in MvcOptions.ValueProviderFactories as the last entry. This means that you'll accept the jQuery syntax in the query string, anywhere in your app.

If the JQuery value provider is being filtered out on `GET` versus `POST` how is it that my own valueProvider would be available? What conditions are checked in order to determine what value providers are used?

Thanks for your help

**Comment by rynowak** [comment]
> If the JQuery value provider is being filtered out on GET versus POST how is it that my own valueProvider would be available

The `JQueryFormValueProvider` is not showing up for you in GET requests because GET requests don't have a body. This value provider reads `form-urlencoded` from the request body.

You need to create another value provider that will do the same thing using the query string.

**Comment by shoe788** [comment]
Gotcha, I'm reading the code again and it's making sense now. 

From an earlier comment...

> We've been discussing this internally and we're likely going to just enable this by default in the future, but this should get you going.

You're saying this type of binding would work in a future release? I think I can get by with posting for now, I just want to keep my options open

**Comment by Eilon** [comment]
We are planning to fix this for the 2.1 release.

**Comment by mkArtakMSFT** [comment]
@kichalla, this one may be fixed too. Can you please review this too? Thanks!

**Comment by rynowak** [comment]
I doubled checked and we never actually did fix this. The linked bug was closed as a duplicate.

**Comment by kichalla** [comment]
@mkArtakMSFT  This looks like a dupe of https://github.com/aspnet/Mvc/issues/6372

**Comment by kichalla** [comment]
Closing as duplicate of https://github.com/aspnet/Mvc/issues/6372
